### PR TITLE
Allowing inference on multiple videos via `sleap-track`

### PR DESCRIPTION
**Pull request by emdavis02**
### Description
Adding an option to add a directory to `sleap-track` to allow the command to iterate through the files in the directory and run inferences on all of them.

### Types of changes

- [ ] Bugfix
- [x] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
#1777 

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [x] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

## Summary by CodeRabbit

- **New Features**
  - Added support for handling multiple video inputs for inference, accommodating various video paths.
  - Inference results now include detailed provenance metadata.
  
- **Improvements**
  - Enhanced error handling for smoother processing of multiple inputs and outputs.
  
- **Bug Fixes**
  - Fixed issues related to running inference on existing predictions and managing output paths.

- **Tests**
  - Introduced new test scenarios for multiple video inputs and output handling.
  - Improved existing tests to cover additional edge cases and invalid inputs.

### CODE REVIEW

File: tests/nn/test_inference.py [hunk 1]
```diff
@@ -1473,6 +1485,155 @@ def test_sleap_track(
     args = [slp_path, "--cpu"]
     with pytest.raises(ValueError):
         sleap_track(args=args)
+        
+@pytest.mark.parametrize("tracking", ["simple", "flow", "None"])
```
Comment (eberrigan) [inline tests/nn/test_inference.py:1489]: add test functions for --output argument where output can be a filename when the input is a single filename and a folder to dump the predictions in when the input is a folder

File: sleap/nn/inference.py [hunk 2]
```diff
@@ -5461,43 +5485,134 @@ def main(args: Optional[list] = None):
...
+
+            # Save results.
+            labels_pr.save(output_path)
+            print("Saved output:", output_path)
+
+            if args.open_in_gui:
+                subprocess.call(["sleap-label", output_path])
+                
+            # Reset output_path for next iteration
+            output_path = args.output
```
Comment (eberrigan) [inline sleap/nn/inference.py:5615]: add output folder as an option when the input is a folder

File: tests/nn/test_inference.py [hunk 3]
```diff
@@ -1473,6 +1485,155 @@ def test_sleap_track(
...
+                new_slp_path_list.append(file_path)
+    elif Path.is_file(slp_path):
+        new_slp_path_list = [args.data_path]
+    
+    for output_path in slp_path_list:
+        assert Path(output_path).exists()
+        
+    #files_to_remove = set(new_slp_path_list) - set(slp_path_list)
+    #for file in files_to_remove:
+    #    file.unlink()
```
Comment (eberrigan) [inline tests/nn/test_inference.py:1636]: add test functions for parsing the CLI when the input is a folder

File: sleap/nn/inference.py [hunk 4]
```diff
@@ -5288,46 +5288,72 @@ def _make_provider_from_cli(args: argparse.Namespace) -> Tuple[Provider, str]:
...
+        
+    # Check for multiple video inputs
+    # Compile file(s) into a list for later itteration
+    if Path.is_dir(data_path):
+        data_path_list = []
+        for file_path in data_path.iterdir():
+            if file_path.is_file():
+                data_path_list.append(file_path)
+    elif Path.is_file(data_path):
+        data_path_list = [args.data_path]
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap/nn/inference.py:5307]: data_path.is_dir() if data_path is already a Path object
data_path.is_file() if data_path is already a Path object

File: sleap/nn/inference.py [hunk 5]
```diff
@@ -5461,43 +5491,164 @@ def main(args: Optional[list] = None):
...
+            if output_path is None:
+                output_path = f"{data_path}.predictions.slp"
+                output_path_obj = Path(output_path)
 
-        labels_pr.provenance["model_paths"] = predictor.model_paths
-        labels_pr.provenance["predictor"] = type(predictor).__name__
+            # if output_path was provided and multiple inputs were provided, create a directory to store outputs
+            elif len(data_path_list) > 1:
+                output_path = (
+                    output_path + "/" + (data_path_obj.stem + ".predictions.slp")
```
Comment (eberrigan) [inline sleap/nn/inference.py:5538]: you have `output_path_obj` defined above so this should be `output_path_obj / (data_path_obj.stem + ".predictions.slp")`

**Review by eberrigan — CHANGES_REQUESTED** [review]
mostly clean up print statements that should not be included in develop branch and make the Path notation consistent / minimize defining the same path variable more than once

**Comment by talmo** [comment]
Review notes from subgroup:
- Remove duplicate files from fixtures and instead just copy them on the fly inside the test into a temp dir
- Update docs to reference the new functionality (at least add usage examples to the CLI doc page for `sleap-track`)

File: sleap/nn/inference.py [hunk 6]
```diff
@@ -5419,11 +5450,9 @@ def main(args: Optional[list] = None):
     # Parse inputs.
     args, _ = parser.parse_known_args(args)
     print("Args:")
-    pprint(vars(args))
+    print(vars(args))
```
Comment (talmo) [inline sleap/nn/inference.py:5453]: Prefer `pprint` (from the `rich` module) for nicely formatted logs.

File: sleap/nn/inference.py [hunk 7]
```diff
@@ -5339,6 +5369,7 @@ def _make_predictor_from_cli(args: argparse.Namespace) -> Predictor:
     Returns:
         The `Predictor` created from loaded models.
     """
+    print(args)
```
Comment (talmo) [inline sleap/nn/inference.py:5372]: Remove

File: sleap/nn/inference.py [hunk 8]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+            data_path_obj = Path(data_path)
+            predictor = _make_predictor_from_cli(args)
+            predictor.tracker = tracker
+
+            # Run inference!
+            labels_pr = predictor.predict(provider)
+
+            # if output path was not provided, create an output path
+            if output_path is not None:
+                print("HERE")
```
Comment (talmo) [inline sleap/nn/inference.py:5518]: Remove

File: sleap/nn/inference.py [hunk 9]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+                output_path_obj = Path(output_path)
+                print(output_path_obj.exists())
+                print(output_path_obj.is_dir())
```
Comment (talmo) [inline sleap/nn/inference.py:5521]: Remove

File: sleap/nn/inference.py [hunk 10]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+                if (
+                    output_path_obj.exists()
+                    and output_path_obj.is_file()
+                    and len(data_path_list) > 1
+                ):
+                    raise ValueError(
+                        "output_path argument must be a directory if multiple video inputs are given"
+                    )
```
Comment (talmo) [inline sleap/nn/inference.py:5529]: This check should occur before any prediction happens. You'll need to have a separate loop over `zip(data_path_list, provider_list)` before this one and apply this check.

This is because some inference jobs will take a really long time, and it'd suck to spend hours of GPU compute and then not have it save anything out.

File: sleap/nn/inference.py [hunk 11]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+            labels_pr.provenance["platform"] = platform.platform()
+            labels_pr.provenance["command"] = " ".join(sys.argv)
+            labels_pr.provenance["data_path"] = data_path_obj.as_posix()
+            labels_pr.provenance["output_path"] = output_path_obj.as_posix()
+            labels_pr.provenance["total_elapsed"] = total_elapsed
+            labels_pr.provenance["start_timestamp"] = start_timestamp
+            labels_pr.provenance["finish_timestamp"] = finish_timestamp
+
+            print("Provenance:")
+            print(labels_pr.provenance)
```
Comment (talmo) [inline sleap/nn/inference.py:5568]: `pprint`

File: sleap/nn/inference.py [hunk 12]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
-        labels_pr.provenance["model_paths"] = predictor.model_paths
-        labels_pr.provenance["predictor"] = type(predictor).__name__
+            # if output_path was provided and multiple inputs were provided, create a directory to store outputs
+            elif len(data_path_list) > 1:
+                output_path = (
+                    output_path_obj.as_posix()
+                    + "/"
+                    + (data_path_obj.stem + ".predictions.slp")
+                )
+                output_path_obj = Path(output_path)
```
Comment (talmo) [inline sleap/nn/inference.py:5542]:
```suggestion
                output_path_obj = Path(output_path)
                
                # Create the containing directory if needed.
                output_path_obj.parent.mkdir(exist_ok=True, parents=True)
```

File: sleap/nn/inference.py [hunk 13]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
 
-        labels_pr.provenance["model_paths"] = predictor.model_paths
-        labels_pr.provenance["predictor"] = type(predictor).__name__
+            # if output_path was provided and multiple inputs were provided, create a directory to store outputs
+            elif len(data_path_list) > 1:
+                output_path = (
+                    output_path_obj.as_posix()
+                    + "/"
+                    + (data_path_obj.stem + ".predictions.slp")
+                )
```
Comment (talmo) [inline sleap/nn/inference.py:5541]: Rather than using string operations, with pathlib you can use the `/` binary operator to concatenate path segments:

```suggestion
                output_path = output_path_obj / data_path_obj.with_suffix(".predictions.slp")
```

File: sleap/nn/inference.py [hunk 14]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
-        print("Starting tracker...")
-        frames = run_tracker(frames=frames, tracker=tracker)
-        tracker.final_pass(frames)
+                elif not output_path_obj.exists() and len(data_path_list) > 1:
+                    output_path = (
+                        output_path_obj.as_posix()
+                        + "/"
+                        + (data_path_obj.stem + ".predictions.slp")
+                    )
+                    output_path_obj = Path(output_path)
```
Comment (talmo) [inline sleap/nn/inference.py:5619]:
```suggestion
                    output_path = output_path_obj / data_path_obj.with_suffix(".predictions.slp")
                    output_path_obj = Path(output_path)
                    output_path_obj.parent.mkdir(exist_ok=True, parents=True)
```

File: sleap/nn/inference.py [hunk 15]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+            labels_pr.provenance["platform"] = platform.platform()
+            labels_pr.provenance["command"] = " ".join(sys.argv)
+            labels_pr.provenance["data_path"] = data_path_obj.as_posix()
+            labels_pr.provenance["output_path"] = output_path_obj.as_posix()
+            labels_pr.provenance["total_elapsed"] = total_elapsed
+            labels_pr.provenance["start_timestamp"] = start_timestamp
+            labels_pr.provenance["finish_timestamp"] = finish_timestamp
+
+            print("Provenance:")
+            print(labels_pr.provenance)
```
Comment (talmo) [inline sleap/nn/inference.py:5642]: `pprint`

File: sleap/nn/inference.py [hunk 16]
```diff
@@ -5461,43 +5490,169 @@ def main(args: Optional[list] = None):
...
+            frames = run_tracker(frames=frames, tracker=tracker)
+            tracker.final_pass(frames)
+
+            labels_pr = Labels(labeled_frames=frames)
+
+            if output_path is None:
+                output_path = f"{data_path}.{tracker.get_name()}.slp"
+                output_path_obj = Path(output_path)
+
+            if output_path is not None:
```
Comment (talmo) [inline sleap/nn/inference.py:5602]: Probably don't need this check right? By this line, `output_path` is never `None` since it's set to a path in the previous couple of lines if it was `None`.

**Review by talmo — CHANGES_REQUESTED** [review]

**Comment by talmo** [comment]
In addition to the changes above, another comment re: supporting inputs that are multiple SLP files:

- We **do** want to support having multiple SLP files that have predicted poses on all frames and that the user wants to retrack only.
- We **do** want to support having multiple SLP files that have suggestion frames and that the user wants to generate predicted poses for, but not track.
- We **do not** want to overwrite SLP files in any case.
- We **really do not** want to mess around with `.pkg.slp` files, particularly not overwrite them. When re-saving labels that had embedded images, they'll get saved without the embedded images and will restore the source videos. If the user had embedded images, it'd be a serious data loss event to have that file get re-saved without the images. In this case, saving to a separate SLP file is the best compromise.

@roomrys am I missing any other dangerzone cases?

**Comment by talmo** [comment]
> In addition to the changes above, another comment re: supporting inputs that are multiple SLP files:
> 
> * We **do** want to support having multiple SLP files that have predicted poses on all frames and that the user wants to retrack only.
> * We **do** want to support having multiple SLP files that have suggestion frames and that the user wants to generate predicted poses for, but not track.
> * We **do not** want to overwrite SLP files in any case.
> * We **really do not** want to mess around with `.pkg.slp` files, particularly not overwrite them. When re-saving labels that had embedded images, they'll get saved without the embedded images and will restore the source videos. If the user had embedded images, it'd be a serious data loss event to have that file get re-saved without the images. In this case, saving to a separate SLP file is the best compromise.
> 
> @roomrys am I missing any other dangerzone cases?

On further assessment during subgroup: let's actually just not do multiple SLP file support.

File: sleap/nn/inference.py [hunk 17]
```diff
@@ -5461,43 +5495,168 @@ def main(args: Optional[list] = None):
...
 
+                # if output_path was provided and multiple inputs were provided, create a directory to store outputs
+                if len(data_path_list) > 1:
+                    output_path = (
+                        output_path_obj
+                        / data_path_obj.with_suffix(".predictions.slp").name
+                    )
+                    output_path_obj = Path(output_path)
+                    # Create the containing directory if needed.
+                    output_path_obj.parent.mkdir(exist_ok=True, parents=True)
```
Comment (eberrigan) [inline sleap/nn/inference.py:5547]: is this what you wanted to do, or did you want to make the directory itself instead of the parent?

**Review by eberrigan — APPROVED** [review]

**Review by talmo — APPROVED** [review]
LGTM!
